### PR TITLE
DAOS-9959 common: verify integer key in the server side

### PR DESCRIPTION
**Pull request by wangshilong**
Signed-off-by: Wang Shilong <shilong.wang@intel.com>

### CODE REVIEW

File: src/common/btree.c [hunk 1]
```diff
@@ -405,6 +405,18 @@ btr_hkey_size(struct btr_context *tcx)
 	return btr_hkey_size_const(btr_ops(tcx), tcx->tc_feats);
 }
 
+static int
+btr_verify_key(struct btr_context *tcx, d_iov_t *key)
```
Comment (jolivier23) [inline src/common/btree.c:409]: inline ?

**Review by jolivier23 — APPROVED** [review]

**Review by jolivier23 — DISMISSED** [review]

File: src/common/btree.c [hunk 2]
```diff
@@ -2010,6 +2035,10 @@ btr_insert(struct btr_context *tcx, d_iov_t *key, d_iov_t *val, d_iov_t *val_out
 	union btr_rec_buf  rec_buf = {0};
 	int		   rc;
 
+	rc = btr_verify_key(tcx, key);
+	if (rc)
+		return rc;
+
```
Comment (NiuYawei) [inline src/common/btree.c:2041]: This looks redundant, it's already verified in btr_upsert().

It looks bit weird to me that the 'btr_verify_key()' calls are sporadically duplicated in both exported functions (dbtree_xxx) and internal functions (btr_xxx), I think the key should be verify in:
- Either all the exported dbtree_xxx() functions, or;
- The internal functions like btr_hkey_gen() and hkey_int_gen();
What do you think?

File: src/common/btree.c [hunk 3]
```diff
@@ -431,7 +431,7 @@ void
 hkey_int_gen(d_iov_t *key,  void *hkey)
 {
 	/* Use key directory as unsigned integer in lieu of hkey */
-	D_ASSERT(key->iov_len <= sizeof(uint64_t));
+	D_ASSERT(key->iov_len == sizeof(uint64_t));
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/common/btree.c:434]: We added same check on the client, so i think it fine here.

File: src/common/btree.c (hunk 2 above)
Comment (wangshilong) [reply src/common/btree.c:2041]: Sure, let's move check earlier to dbtree_xxxx()

**Review by NiuYawei — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**